src/attributes.jl

```diff
@@ -461,6 +517,136 @@ struct RawParameter <: AbstractOptimizerAttribute
+to you to provide the appropriate one. If in doubt, give a complete solution.
+
+The value of the attribute should be a function taking only one argument, that
+is commonly called `callback_data`, that can be used for instance in
+[`LazyConstraintCallback`](@ref), [`HeuristicCallback`](@ref) and
+[`UserCutCallback`](@ref).
+"""
+abstract type AbstractCallback <: AbstractOptimizerAttribute end
+
+"""
+    LazyConstraintCallback() <: AbstractCallback
+
+The callback can be used to reduce the feasible set given the current primal
+solution by submitting a [`LazyConstraint`](@ref). For instance, it may be
+called at an incumbent of a mixed-integer problem. Note that there is no
+guarantee that the callback is called at *every* feasible primal solution.
+
+The feasible primal solution is accessed through
+[`CallbackVariablePrimal`](@ref). Trying to access other result
+attributes will throw [`OptimizeInProgress`](@ref) as discussed in
+[`AbstractCallback`](@ref).
+
+## Examples
+
+```julia
+x = MOI.add_variables(optimizer, 8)
+MOI.submit(optimizer, MOI.LazyConstraintCallback(), callback_data -> begin
+    sol = MOI.get(optimizer, MOI.CallbackVariablePrimal(callback_data), x)
+    if # should add a lazy constraint
+        func = # computes function
+        set = # computes set
+        MOI.submit(optimizer, MOI.LazyConstraint(callback_data), func, set)
+    end
+end
+```
+"""
+struct LazyConstraintCallback <: AbstractCallback end
+
+"""
+    HeuristicCallback() <: AbstractCallback
+
+The callback can be used to submit [`HeuristicSolution`](@ref) given the
+current primal solution.
+For instance, it may be called at fractional (i.e., non-integer) nodes in the
+branch and bound tree of a mixed-integer problem. Note that there is not
+guarantee that the callback is called *everytime* the solver has an infeasible
+solution.
+
+The infeasible solution is accessed through
+[`CallbackVariablePrimal`](@ref). Trying to access other result
+attributes will throw [`OptimizeInProgress`](@ref) as discussed in
+[`AbstractCallback`](@ref).
+
+## Examples
+
+```julia
+x = MOI.add_variables(optimizer, 8)
+MOI.submit(optimizer, MOI.HeuristicCallback(), callback_data -> begin
+    sol = MOI.get(optimizer, MOI.CallbackVariablePrimal(callback_data), x)
+    if # can find a heuristic solution
+        values = # computes heuristic solution
+        MOI.submit(optimizer, MOI.HeuristicSolution(callback_data), x,
+                   values)
+    end
+end
+```
+"""
+struct HeuristicCallback <: AbstractCallback end
+
+"""
+    UserCutCallback() <: AbstractCallback
+
+The callback can be used to submit [`UserCut`](@ref) given the current primal
+solution. For instance, it may be called at fractional (i.e., non-integer) nodes
+in the branch and bound tree of a mixed-integer problem. Note that there is not
+guarantee that the callback is called *everytime* the solver has an infeasible
+solution.
+
+The infeasible solution is accessed through
```

Comment:
> infeasible solution

This terminology might be confusing in MILP context, since. "Fractional solution" would be less ambiguous.

Reply:
Let's stay general and say "current solution" to cover the case of heuristic callback from incumbent.

Reply:
Would it be better to use "solution candidate", then, because it's not known whether the values are feasible, and can be a solution? In terms of B&B tree nodes, one could also talk of current relaxation, or current relaxation's solution.